File: library/src/test/java/com/deltadna/android/sdk/net/NetworkDispatcherTest.java
/*
 * Copyright (c) 2016 deltaDNA Ltd. All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.deltadna.android.sdk.net;

import android.os.Handler;

import com.deltadna.android.sdk.exceptions.ResponseException;
import com.deltadna.android.sdk.listeners.RequestListener;
import com.squareup.okhttp.mockwebserver.MockResponse;
import com.squareup.okhttp.mockwebserver.MockWebServer;

import org.junit.After;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentMatcher;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.runners.MockitoJUnitRunner;
import org.mockito.stubbing.Answer;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.argThat;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public final class NetworkDispatcherTest {
    
    private static final int MAX_RETRIES = 1;
    
    private NetworkDispatcher uut;
    private MockWebServer server;
    
    @Before
    public void before() throws IOException {
        final Handler handler = mock(Handler.class);
        when(handler.post(any(Runnable.class))).then(new Answer<Void>() {
            @Override
            public Void answer(final InvocationOnMock invocation) throws Throwable {
                ((Runnable) invocation.getArguments()[0]).run();
                return null;
            }
        });
        
        uut = new NetworkDispatcher(handler);
        
        server = new MockWebServer();
        server.start();
    }
    
    @After
    public void after() throws IOException {
        server.shutdown();
        uut = null;
    }
    
    @Test
    public void successfulRequest() throws InterruptedException {
        final String responseBody = "response";
        final RequestListener<Response<String>> listener =
                mock(RequestListener.class);
        
        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setBody(responseBody));
        
        uut.enqueue(
                new Request.Builder<String>()
                        .get()
                        .url(server.url("/success").toString())
                        .build(),
                ResponseBodyConverter.STRING,
                listener);
        
        server.takeRequest();
        Thread.sleep(100);
        
        assertThat(server.getRequestCount()).isEqualTo(1);
        verify(listener).onSuccess(eq(new Response<String>(
                200, responseBody.getBytes(), responseBody)));
    }
    
    @Test
    public void failureRetriesRequest()
            throws InterruptedException {
        
        final String responseBody = "not found";
        final RequestListener<Response<Void>> listener =
                mock(RequestListener.class);
        
        for (int i = 0; i < 1 + MAX_RETRIES; i++) {
            server.enqueue(new MockResponse()
                    .setResponseCode(404)
                    .setBody(responseBody));
        }
        
        uut.enqueue(
                new Request.Builder<Void>()
                        .get()
                        .url(server.url("/failure").toString())
                        .maxRetries(MAX_RETRIES)
                        .build(),
                listener);
        
        for (int i = 0; i < 1 + MAX_RETRIES; i++) {
            server.takeRequest();
        }
        Thread.sleep(100);
        
        assertThat(server.getRequestCount()).isEqualTo(MAX_RETRIES + 1);
        verify(listener).onFailure(argThat(new ArgumentMatcher<Throwable>() {
            @Override
            public boolean matches(Object argument) {
                assertThat(argument).isInstanceOf(ResponseException.class);
                assertThat(((ResponseException) argument).response)
                        .isEqualTo(new Response<String>(
                                404,
                                responseBody.getBytes(),
                                responseBody));
                return true;
            }
        }));
    }
    
    @Test
    public void retriesRespectDelay() throws InterruptedException {
        final RequestListener<Response<Void>> listener =
                mock(RequestListener.class);
        
        server.enqueue(new MockResponse().setResponseCode(404));
        server.enqueue(new MockResponse().setResponseCode(200));
        
        uut.enqueue(
                new Request.Builder<Void>()
                        .get()
                        .url(server.url("/delay").toString())
                        .maxRetries(1)
                        .retryDelay(1000)
                        .build(),
                listener);
        
        server.takeRequest();
        final long first = System.currentTimeMillis();
        server.takeRequest();
        final long second = System.currentTimeMillis();
        Thread.sleep(100);
        
        assertThat(second - first).isGreaterThan(900L);
        assertThat(second - first).isAtMost(1100L);
        assertThat(server.getRequestCount()).isEqualTo(2);
        verify(listener).onSuccess(any(Response.class));
    }
    
    // FIXME test failing only on jenkins
    @Ignore
    @Test
    public void requestCancellation() throws InterruptedException {
        final RequestListener<Response<Void>> listener =
                mock(RequestListener.class);
        
        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setBodyDelay(200, TimeUnit.MILLISECONDS));
        uut.enqueue(
                new Request.Builder<Void>()
                        .get()
                        .url(server.url("/cancel").toString())
                        .build(),
                listener)
                .cancel();
        server.takeRequest(100, TimeUnit.MILLISECONDS);
        Thread.sleep(500);
        
        assertThat(server.getRequestCount()).isEqualTo(1);
        verifyZeroInteractions(listener);
    }
}

File: library/src/test/java/com/deltadna/android/sdk/net/NetworkDispatcherTest.kt
/*
 * Copyright (c) 2016 deltaDNA Ltd. All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.deltadna.android.sdk.net

import com.deltadna.android.sdk.listeners.RequestListener
import com.google.common.truth.Truth.assertThat
import com.nhaarman.mockito_kotlin.*
import com.squareup.okhttp.mockwebserver.MockResponse
import com.squareup.okhttp.mockwebserver.MockWebServer
import org.junit.After
import org.junit.Before
import org.junit.Ignore
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment
import java.io.IOException
import java.util.concurrent.TimeUnit

@RunWith(RobolectricTestRunner::class)
class NetworkDispatcherTest {
    
    private var uut: NetworkDispatcher? = null
    private var server: MockWebServer? = null
    
    @Before
    fun before() {
        uut = NetworkDispatcher()
        
        server = MockWebServer()
        server!!.start()
    }
    
    @After
    fun after() {
        server!!.shutdown()
        server = null
        
        uut = null
    }
    
    @Test
    fun successfulRequest() {
        val responseBody = "response"
        val listener = mock<RequestListener<String>>()
        
        server!!.enqueue(MockResponse()
                .setResponseCode(200)
                .setBody(responseBody))
        
        uut!!.enqueue(
                Request.Builder<String>()
                        .get()
                        .url(server!!.url("/success").toString())
                        .build(),
                ResponseBodyConverter.STRING,
                listener)
        
        server!!.takeRequest()
        Thread.sleep(100)
        RuntimeEnvironment.getMasterScheduler().advanceToNextPostedRunnable()
        
        assertThat(server!!.requestCount).isEqualTo(1)
        verify(listener).onCompleted(eq(Response(
                200,
                responseBody.toByteArray(),
                responseBody,
                null)))
    }
    
    @Test
    fun failureRetriesRequest() {
        val responseBody = "not found"
        val listener = mock<RequestListener<Void>>()
        
        for (i in 0..MAX_RETRIES) {
            server!!.enqueue(MockResponse()
                    .setResponseCode(404)
                    .setBody(responseBody)
                    .setBodyDelay(150, TimeUnit.MILLISECONDS))
        }
        
        uut!!.enqueue(
                Request.Builder<Void>()
                        .get()
                        .url(server!!.url("/failure").toString())
                        .readTimeout(100)
                        .maxRetries(MAX_RETRIES)
                        .build(),
                listener)
        
        Thread.sleep(200L * MAX_RETRIES)
        for (i in 0..MAX_RETRIES) {
            server!!.takeRequest()
        }
        Thread.sleep(100)
        RuntimeEnvironment.getMasterScheduler().advanceToLastPostedRunnable()
        
        assertThat(server!!.requestCount).isEqualTo(MAX_RETRIES + 1)
        verify(listener).onError(isA<IOException>())
    }
    
    @Test
    fun retriesRespectDelay() {
        val listener = mock<RequestListener<Void>>()
        
        server!!.enqueue(MockResponse()
                .setResponseCode(404)
                .setBodyDelay(150, TimeUnit.MILLISECONDS))
        server!!.enqueue(MockResponse()
                .setResponseCode(200)
                .setBodyDelay(150, TimeUnit.MILLISECONDS))
        
        uut!!.enqueue(
                Request.Builder<Void>()
                        .get()
                        .url(server!!.url("/delay").toString())
                        .readTimeout(100)
                        .maxRetries(1)
                        .retryDelay(1000)
                        .build(),
                listener)
        
        server!!.takeRequest()
        val first = System.currentTimeMillis()
        server!!.takeRequest()
        val second = System.currentTimeMillis()
        Thread.sleep(100)
        RuntimeEnvironment.getMasterScheduler().advanceToLastPostedRunnable()
        
        assertThat(second - first).isGreaterThan(900L)
        assertThat(second - first).isAtMost(1100L)
        assertThat(server!!.requestCount).isEqualTo(2)
        verify(listener).onCompleted(any())
    }
    
    // FIXME test failing only on jenkins
    @Ignore
    @Test
    fun requestCancellation() {
        val listener = mock<RequestListener<Void>>()
        
        server!!.enqueue(MockResponse()
                .setResponseCode(200)
                .setBodyDelay(200, TimeUnit.MILLISECONDS))
        uut!!.enqueue(
                Request.Builder<Void>()
                        .get()
                        .url(server!!.url("/cancel").toString())
                        .build(),
                listener).cancel()
        server!!.takeRequest(100, TimeUnit.MILLISECONDS)
        Thread.sleep(500)
        
        assertThat(server!!.requestCount).isEqualTo(1)
        assertThat(RuntimeEnvironment.getMasterScheduler().areAnyRunnable())
                .isFalse()
        verifyZeroInteractions(listener)
    }
    
    companion object {
        
        private val MAX_RETRIES = 1
    }
}
